OneSE: the One-Standard-Error Rule

dPys opened the pull request:
#### Reference Issues/PRs
https://github.com/scikit-learn/scikit-learn/blob/master/examples/model_selection/plot_grid_search_refit_callable.py
See also #11269. See also #11354. See also #12865. See also #9499.

#### What does this implement/fix? Explain your changes.
As discussed briefly with @amueller and @NicolasHug last week, this is an enhancement to the latest implementation of sklearn's refit callable functionality. The aim is to provide a more generic set of methods for balancing model complexity with CV performance via model selection 'smoothing'.  In the context of a highly versatile package such as sklearn, where it becomes possible to fit the vast majority of model types using an extensive set of parameters and scorers, the risk of overfitting may be more pressing. To ameliorate this, OneSE might be especially valuable for final model estimation in CV.

#### Any other comments?
The challenge with implementing such a tool has always been with generalizing its functionality to all types of models, parameters, and scoring methods. In particular, defining model 'complexity' is also relatively context-dependent.  This PR lays out a prototype that allows these definitions to be user-determined, rather than hard-coded defaults. Currently, both _error and _score type scorers are supported, along with multi-metric scoring. Either 1 SD bounds can be used, or a percentile tolerance can be specified. Feel free to revise, rewrite, and discuss!

Expanding upon the excellent example recently created by @jiaowoshabi :
```py
import numpy as np
import matplotlib.pyplot as plt
from sklearn.datasets import load_digits
from sklearn.decomposition import PCA
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection._search import OneSE
from sklearn.pipeline import Pipeline
from sklearn.svm import LinearSVC

pipe = Pipeline([
        ('reduce_dim', PCA(random_state=42)),
        ('classify', LinearSVC(random_state=42)),
])

param_grid = {
    'reduce_dim__n_components': [2, 4, 6, 8]
}

# Here, we set our OneSE parameters, though ideally, we would make them explicit arguments for the actual OneSE callable below:
param='n_components'
greater_is_complex=True
refit_scoring = 'accuracy'
tol = None

grid = GridSearchCV(pipe, cv=10, n_jobs=1, param_grid=param_grid,
                    scoring=['accuracy', 'neg_mean_squared_log_error'], refit=OneSE)
digits = load_digits()
grid.fit(digits.data, digits.target)

n_components = grid.cv_results_['param_reduce_dim__n_components']
test_scores = grid.cv_results_['mean_test_accuracy']

plt.figure()
plt.bar(n_components, test_scores, width=1.3, color='b')

plt.axhline(np.max(test_scores), linestyle='--', color='y', label='Best score')

plt.title("Balance model complexity and cross-validated score")
plt.xlabel('Number of PCA components used')
plt.ylabel('Digit classification accuracy')
plt.xticks(n_components.tolist())
plt.ylim((0, 1.0))
plt.legend(loc='upper left')

best_index_ = grid.best_index_

print("The best_index_ is %d" % best_index_)
print("The n_components selected is %d" % n_components[best_index_])
print("The corresponding accuracy score is %.2f"
      % grid.cv_results_['mean_test_accuracy'][best_index_])
plt.show()
```

Let me know what you think
@dPys

jayybhatt committed:
Added validation test for iforest on uniform data (#14771)

OneSE

OneSE - fix typo in docstring

OneSE - add function partial for adding explicit user arguments to callable

Change name to SmoothCV, add support for Wilcoxon method, reorganize callable partial to query the new SmoothCV class.

rename to RazorCV, create routine-independent callables

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -1484,3 +1484,203 @@ def _run_search(self, evaluate_candidates):
+
+
+
+def OneSE(cv_results, **kwargs):
```
We avoid **kw. Please define parameters explicitly. This is explained in our developers' guide

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -1484,3 +1484,203 @@ def _run_search(self, evaluate_candidates):
+    cv_results : dict of numpy(masked) ndarrays
+        See attribute cv_results_ of `GridSearchCV`.
+    param : str
+        Parameter with the larges influence on model complexity.
```
```suggestion
        Parameter with the largest influence on model complexity.
```

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -1484,3 +1484,203 @@ def _run_search(self, evaluate_candidates):
+        Whether complexity increases as `param` increases. Default is True.
+    refit_scoring : str
+        Scoring metric.
+    tol : float
```
Doesn't this effectively make it not 1se?

jnothman reviewed (commented):
Thanks. Yes this might be a good thing to have, but it's not clear why we should support 1se rather than a formal hypothesis test.

 This would need tests and user guide / examples

jnothman commented:
I'm still not sure why we should prefer 1se over a hypothesis test like
wilcoxon.

Of course this isn't super flexible in terms of how it measures model
complexity. I think that's okay.

jnothman commented:
As far as I understand, 1se is rule of thumb approach to approximate
insignificant difference. But there are more precise statistical,
nonparametric approaches to determining if one system is significantly
better than another.

jnothman commented:
No I think you're looking at samples in the wrong direction. I understand
we are looking for "the least complex model whose performance is
insignificantly different from the best". So you compare each model only to
the best, where the number of samples is the number of test evaluations,
i.e. number of splits.

jnothman commented:
The point is that the scores are paired. You want to see whether there is a
consistent improvement rather than an improvement in the average.

Yes, N is small. That's true for calculating the standard deviation too.
Wilcoxon should be possible for small N though it will not reject the null
hypothesis easily... Which is sort of the point. Feel free to empirically
compare the two with different data sets, or to look for literature to
substantiate one approach over the other.

jnothman commented:
Rather than using a function with the user applying partial, why not define
a class whose instance can be called as refit?

jnothman commented:
Also essential is the metric and the test set variability

On Wed., 28 Aug. 2019, 10:10 am Derek Pisner, <notifications@github.com>
wrote:

> The point is that the scores are paired. You want to see whether there is
> a consistent improvement rather than an improvement in the average. Yes, N
> is small. That's true for calculating the standard deviation too. Wilcoxon
> should be possible for small N though it will not reject the null
> hypothesis easily... Which is sort of the point. Feel free to empirically
> compare the two with different data sets, or to look for literature to
> substantiate one approach over the other.
>
> Another thing to consider is that 1 SE was created in the context of
> decision trees, where 1 SE was empirically found to be optimal for
> balancing tree size with *error* (hence 1-SE vs. 1-SD). Outside of that
> context, 1 SD may not be empirically optimal as a hard and fast rule.
> Evaluating 1-SE vs. a hypothesis test across datasets may yield totally
> different results depending on the type of learner and target parameter
> used. It seems to me that the best workaround is to simply include each of
> these methods as options and let the user decide what is most appropriate
> for the problem at hand?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/14820?email_source=notifications&email_token=AAATH25EZSRURFX455YN4BDQGW657A5CNFSM4IP5KIY2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5JPLIA#issuecomment-525530528>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAATH26QT2UUB3YS3DERCALQGW657ANCNFSM4IP5KIYQ>
> .
>

jnothman commented:
I don't see how metric-model pairing affects things... or class imbalance.
The minimum assumption we can rely on is that a greater metric is better
for a given test set (regardless of model, dataset, CV or metric choice).
In many specific cases we might be able to assume a roughly normal
distribution of scores given a model class.

dPys commented:
@jnothman -- I've updated the PR extensively. These routines have now been renamed to SmoothCV, which reflects a generalization of the OneSE approach for overfit prevention. I've added support for the Wilcoxon method of hyp-testing across folds, and reorganized the callable partial to query the new SmoothCV class, within which all of these routines are now included. This is still very much a WIP, but for now, please let me know if the Wilcoxon approach was along the lines of what you were thinking. Please also feel free to revise with any of your own edits.

Cheers,
@dPys

dPys commented:
And use looks something like this:
```
pipe = Pipeline([
        ('reduce_dim', PCA(random_state=42)),
        ('classify', LinearSVC(random_state=42)),
])

param_grid = {
    'reduce_dim__n_components': [2, 4, 6, 8]
}

param='n_components'
greater_is_complex=True
refit_scoring = 'neg_mean_squared_log_error'
tol = 0.25

grid = GridSearchCV(pipe, cv=10, n_jobs=1, param_grid=param_grid,
                    scoring=['accuracy', 'neg_mean_squared_log_error'], refit=smooth_cv(param, greater_is_complex, refit_scoring, tol))
digits = load_digits()
grid.fit(digits.data, digits.target)
```

jnothman commented:
I don't think smooth_cv is a very clear name. What we are trying to say is
"refit the simplest model that is insignificantly different from the best,
where simplicity is determined from a numeric hyperparameter".

For a name I might rather simplest, or simplest_by_param, or simplest_best,
or razor.

I'm not convinced by combining multiple methods into one name either. I
would be happy if the user had to say:

refit=razors.standard_error(param, scoring='roc_auc')

refit=razors.rank_sum_test(param, scoring='roc_auc', alpha=...)

I think this is much more readable than your current proposal.

jnothman commented:
I'm referring to Ockham's razor btw

dPys commented:
@jnothman -- I agree completely and will make those changes asap.

Stay tuned,
@dPys

dPys commented:
@jnothman 

Squashed/rebased changes.  

Improvements:
*All of the routines are now encapsulated in a class called RazorCV
*Several hours were devoted to rethinking how to allow for more readable callables as suggested-- this is actually a rather difficult problem to solve cleanly given the use of class-embedded function partials as callables in this context. 

The current solution works, but there may be other ways to implement it even more cleanly-- e.g. splitting things up into two classes (one containing the universal routines, and the other that contains each callable method, drawing from the first class). Before going down that road, however, I wanted to get your thoughts since there are many ways to go about this.

Either way, I think we've got a nice WIP here.
@dPys

dPys commented:
```
import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import load_digits
from sklearn.decomposition import PCA
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.svm import LinearSVC
pipe = Pipeline([
        ('reduce_dim', PCA(random_state=42)),
        ('classify', LinearSVC(random_state=42)),
])

param_grid = {
    'reduce_dim__n_components': [2, 4, 6, 8]
}

param='n_components'
greater_is_complex=True
scoring = 'neg_mean_squared_log_error'
alpha = 0.05

grid = GridSearchCV(pipe, cv=10, n_jobs=1, param_grid=param_grid,
                    scoring=['accuracy', 'neg_mean_squared_log_error'], refit=RazorCV.ranksum(param, greater_is_complex, scoring, alpha=alpha))
digits = load_digits()
grid.fit(digits.data, digits.target)
```

jnothman commented:
Please avoid squashing and force pushing. We can squash upon merge but in the meantime it makes a mess of the commit history.

dPys commented:
> Please avoid squashing and force pushing. We can squash upon merge but in the meantime it makes a mess of the commit history.

Oh alright, sorry about that. I see now that the push caught #14771 by mistake.

dPys commented:
Assuming we are at least mostly happy with the current design, it seems like a good next step would be to fuzz test functionality with more types of learners, scoring methods, and feature-selecting parameters (i.e. beyond PCA)? ...maybe even create a parameterized test across various combinations of available options currently offered with scikit-learn?  Do you know of any potential volunteers who'd be interested in assisting with this?

Thanks for the ongoing help and encouragement :-)
@dPys
